### ISSUE TITLE
Add URI and Path backtick template types

### ISSUE THREAD

**Issue by sanjiva** [ballerina-platform/ballerina-spec#451]
We currently use string for all URIs and URI paths. If we had template types for these we could tighten static checking for listeners and remote method invocations.

Note that the type should be in the signature so this becomes natural:
```ballerina
public type Client object {
   new (uri host) {
   }
   remote function get (uripath path) returns http:Response|error {
   }
}
```
Example:
```ballerina
http:Client hc = new ("http://foo.com");
http:Response hr = hc->get ("/bar?x=10");
hr = hc->get ("/bar?x=10/y"); // will get a syntax error
```

This should build on string templates so its really a string template + format. This is effectively a few useful special cases of string templating.

Ideally can we generalize this so a new template type can be added by users even? They'd need to implement code to validate, parser and all that. This is a compiler extension of course.

**Comment by jclark** [comment]
I see this as more to do with #239:

```
const URI_PATTERN = "regexp for URIs";

@constrain { pattern: URI_PATTERN }
type Uri string;
```

Backtick templates only help if you want to automatically percent-encode significant characters within a substitution.

**Comment by sanjiva** [comment]
Yes this particular case could be handled with a pattern constraint on strings. Using backtick templates to do percent encoding is quite helpful actually. Right now we do that inside the function, which is not right if the type is a URI.

The tagged templates are a very powerful concept, especially now with the raw backtick concept we're adding in #442. It would be great to take it a step further and have a way to register a compile time handler for a particular backtick type (e.g. xml, html) so that that code can validate the expressions etc.. The runtime validation could also be done similarly.
